File: go/squiggle.go
package main

import "fmt"
import "math"
import "sync"
import rand "math/rand/v2"

type src = *rand.Rand
type func64 = func(src) float64

// https://pkg.go.dev/math/rand/v2

func sample_unit_uniform(r src) float64 {
	return r.Float64()
}

func sample_unit_normal(r src) float64 {
	return r.NormFloat64()
}

func sample_uniform(start float64, end float64, r src) float64 {
	return sample_unit_uniform(r)*(end-start) + start
}

func sample_normal(mean float64, sigma float64, r src) float64 {
	return mean + sample_unit_normal(r)*sigma
}

func sample_lognormal(logmean float64, logstd float64, r src) float64 {
	return (math.Exp(sample_normal(logmean, logstd, r)))
}

func sample_normal_from_90_ci(low float64, high float64, r src) float64 {
	var normal90 float64 = 1.6448536269514727
	var mean float64 = (high + low) / 2.0
	var std float64 = (high - low) / (2.0 * normal90)
	return sample_normal(mean, std, r)

}

func sample_to(low float64, high float64, r src) float64 {
	// Given a (positive) 90% confidence interval,
	// returns a sample from a lognorma with a matching 90% c.i.
	// Key idea: If we want a lognormal with 90% confidence interval [a, b]
	// we need but get a normal with 90% confidence interval [log(a), log(b)].
	// Then see code for sample_normal_from_90_ci
	var loglow float64 = math.Log(low)
	var loghigh float64 = math.Log(high)
	return math.Exp(sample_normal_from_90_ci(loglow, loghigh, r))
}

func sample_mixture(fs []func64, weights []float64, r src) float64 {

	// fmt.Println("weights initially: ", weights)
	var sum_weights float64 = 0
	for _, weight := range weights {
		sum_weights += weight
	}

	var total float64 = 0
	var cumsummed_normalized_weights = append([]float64(nil), weights...)
	for i, weight := range weights {
		total += weight / sum_weights
		cumsummed_normalized_weights[i] = total
	}

	var result float64
	var flag int = 0
	var p float64 = r.Float64()

	for i, cnw := range cumsummed_normalized_weights {
		if p < cnw {
			result = fs[i](r)
			flag = 1
			break
		}
	}
	// fmt.Println(cumsummed_normalized_weights)

	if flag == 0 {
		result = fs[len(fs)-1](r)
	}
	return result

}

func slice_fill(xs []float64, fs func64, r src) {
	for i := range xs {
		xs[i] = fs(r)
	}
}

func sample_parallel(f func64, n_samples int) []float64 {
	var num_threads = 16
	var xs = make([]float64, n_samples)
	var wg sync.WaitGroup
	var h = n_samples / num_threads
	wg.Add(num_threads)
	for i := range num_threads {
		var xs_i = xs[i*h : (i+1)*h]
		go func(f func64) {
			defer wg.Done()
			var r = rand.New(rand.NewPCG(uint64(i), uint64(i+1)))
			for i := range xs_i {
				xs_i[i] = f(r)
			}
		}(f)
	}

	wg.Wait()
	return xs

}

func main() {

	var p_a float64 = 0.8
	var p_b float64 = 0.5
	var p_c float64 = p_a * p_b
	ws := [4](float64){1 - p_c, p_c / 2, p_c / 4, p_c / 4}

	sample_0 := func(r src) float64 { return 0 }
	sample_1 := func(r src) float64 { return 1 }
	sample_few := func(r src) float64 { return sample_to(1, 3, r) }
	sample_many := func(r src) float64 { return sample_to(2, 10, r) }
	fs := [4](func64){sample_0, sample_1, sample_few, sample_many}

	model := func(r src) float64 { return sample_mixture(fs[0:], ws[0:], r) }
	n_samples := 1_000_000
	xs := sample_parallel(model, n_samples)
	var avg float64 = 0
	for _, x := range xs {
		avg += x
	}
	avg = avg / float64(n_samples)
	fmt.Printf("Average: %v\n", avg)
	/*
		n_samples := 1_000_000
		var r = rand.New(rand.NewPCG(uint64(1), uint64(2)))
		var avg float64 = 0
		for i := 0; i < n_samples; i++ {
			avg += sample_mixture(fs[0:], ws[0:], r)
		}
		avg = avg / float64(n_samples)
		fmt.Printf("Average: %v\n", avg)
	*/
}

File: squiggle.c/squiggle_c/squiggle.c
#include <limits.h>
#include <math.h>
#include <stdint.h>
#include <stdlib.h>

// Defs
#define PI 3.14159265358979323846 // M_PI in gcc gnu99
#define NORMAL90CONFIDENCE 1.6448536269514727
#define UNUSED(x) (void)(x)
// ^ https://stackoverflow.com/questions/3599160/how-can-i-suppress-unused-parameter-warnings-in-c

// Pseudo Random number generators
static uint64_t xorshift64(uint64_t* seed)
{
    // Algorithm "xor" from p. 4 of Marsaglia, "Xorshift RNGs"
    // See:
    //   <https://en.wikipedia.org/wiki/Xorshift>
    //   <https://stackoverflow.com/questions/53886131/how-does-xorshift32-works>,
    // Also some drama:
    //   <https://www.pcg-random.org/posts/on-vignas-pcg-critique.html>,
    //   <https://prng.di.unimi.it/>
    uint64_t x = *seed;
    x ^= x << 13;
    x ^= x >> 7;
    x ^= x << 17;
    return *seed = x;

    /* 
    // if one wanted to generate 32 bit ints, 
    // from which to generate floats,
    // one could do the following: 
    uint32_t x = *seed;
    x ^= x << 13;
    x ^= x >> 17;
    x ^= x << 5;
    return *seed = x;
    */
}

// Distribution & sampling functions
// Unit distributions
double sample_unit_uniform(uint64_t* seed)
{
    // samples uniform from [0,1] interval.
    return ((double)xorshift64(seed)) / ((double)UINT64_MAX);
}

double sample_unit_normal(uint64_t* seed)
{
    // // See: <https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform>
    double u1 = sample_unit_uniform(seed);
    double u2 = sample_unit_uniform(seed);
    double z = sqrt(-2.0 * log(u1)) * sin(2 * PI * u2);
    return z;
}

// Composite distributions
double sample_uniform(double start, double end, uint64_t* seed)
{
    return sample_unit_uniform(seed) * (end - start) + start;
}

double sample_normal(double mean, double sigma, uint64_t* seed)
{
    return (mean + sigma * sample_unit_normal(seed));
}

double sample_lognormal(double logmean, double logstd, uint64_t* seed)
{
    return exp(sample_normal(logmean, logstd, seed));
}

double sample_normal_from_90_ci(double low, double high, uint64_t* seed)
{
    // Explanation of key idea:
    // 1. We know that the 90% confidence interval of the unit normal is
    // [-1.6448536269514722, 1.6448536269514722]
    // see e.g.: https://stackoverflow.com/questions/20626994/how-to-calculate-the-inverse-of-the-normal-cumulative-distribution-function-in-p
    // or https://www.wolframalpha.com/input?i=N%5BInverseCDF%28normal%280%2C1%29%2C+0.05%29%2C%7B%E2%88%9E%2C100%7D%5D
    // 2. So if we take a unit normal and multiply it by
    // L / 1.6448536269514722, its new 90% confidence interval will be
    // [-L, L], i.e., length 2 * L
    // 3. Instead, if we want to get a confidence interval of length L,
    // we should multiply the unit normal by
    // L / (2 * 1.6448536269514722)
    // Meaning that its standard deviation should be multiplied by that amount
    // see: https://en.wikipedia.org/wiki/Normal_distribution?lang=en#Operations_on_a_single_normal_variable
    // 4. So we have learnt that Normal(0, L / (2 * 1.6448536269514722))
    // has a 90% confidence interval of length L
    // 5. If we want a 90% confidence interval from high to low,
    // we can set mean = (high + low)/2; the midpoint, and L = high-low,
    // Normal([high + low]/2, [high - low]/(2 * 1.6448536269514722))
    double mean = (high + low) / 2.0;
    double std = (high - low) / (2.0 * NORMAL90CONFIDENCE);
    return sample_normal(mean, std, seed);
}

double sample_to(double low, double high, uint64_t* seed)
{
    // Given a (positive) 90% confidence interval,
    // returns a sample from a lognorma with a matching 90% c.i.
    // Key idea: If we want a lognormal with 90% confidence interval [a, b]
    // we need but get a normal with 90% confidence interval [log(a), log(b)].
    // Then see code for sample_normal_from_90_ci
    double loglow = log(low);
    double loghigh = log(high);
    return exp(sample_normal_from_90_ci(loglow, loghigh, seed));
}

double sample_gamma(double alpha, uint64_t* seed)
{

    // A Simple Method for Generating Gamma Variables, Marsaglia and Wan Tsang, 2001
    // https://dl.acm.org/doi/pdf/10.1145/358407.358414
    // see also the references/ folder
    // Note that the Wikipedia page for the gamma distribution includes a scaling parameter
    // k or beta
    // https://en.wikipedia.org/wiki/Gamma_distribution
    // such that gamma_k(alpha, k) = k * gamma(alpha)
    // or gamma_beta(alpha, beta) = gamma(alpha) / beta
    // So far I have not needed to use this, and thus the second parameter is by default 1.
    if (alpha >= 1) {
        double d, c, x, v, u;
        d = alpha - 1.0 / 3.0;
        c = 1.0 / sqrt(9.0 * d);
        while (1) {

            do {
                x = sample_unit_normal(seed);
                v = 1.0 + c * x;
            } while (v <= 0.0);

            v = v * v * v;
            u = sample_unit_uniform(seed);
            if (u < 1.0 - 0.0331 * (x * x * x * x)) { // Condition 1
                // the 0.0331 doesn't inspire much confidence
                // however, this isn't the whole story
                // by knowing that Condition 1 implies condition 2
                // we realize that this is just a way of making the algorithm faster
                // i.e., of not using the logarithms
                return d * v;
            }
            if (log(u) < 0.5 * (x * x) + d * (1.0 - v + log(v))) { // Condition 2
                return d * v;
            }
        }
    } else {
        return sample_gamma(1 + alpha, seed) * pow(sample_unit_uniform(seed), 1 / alpha);
        // see note in p. 371 of https://dl.acm.org/doi/pdf/10.1145/358407.358414
    }
}

double sample_beta(double a, double b, uint64_t* seed)
{
    // See: https://en.wikipedia.org/wiki/Gamma_distribution#Related_distributions
    double gamma_a = sample_gamma(a, seed);
    double gamma_b = sample_gamma(b, seed);
    return gamma_a / (gamma_a + gamma_b);
}

double sample_laplace(double successes, double failures, uint64_t* seed)
{
    // see <https://en.wikipedia.org/wiki/Beta_distribution?lang=en#Rule_of_succession>
    return sample_beta(successes + 1, failures + 1, seed);
}

// Array helpers
double array_sum(double* array, int length)
{
    double sum = 0.0;
    for (int i = 0; i < length; i++) {
        sum += array[i];
    }
    return sum;
}

void array_cumsum(double* array_to_sum, double* array_cumsummed, int length)
{
    array_cumsummed[0] = array_to_sum[0];
    for (int i = 1; i < length; i++) {
        array_cumsummed[i] = array_cumsummed[i - 1] + array_to_sum[i];
    }
}

double array_mean(double* array, int length)
{
    double sum = array_sum(array, length);
    return sum / length;
}

double array_std(double* array, int length)
{
    double mean = array_mean(array, length);
    double std = 0.0;
    for (int i = 0; i < length; i++) {
        std += (array[i] - mean) * (array[i] - mean);
    }
    std = sqrt(std / length);
    return std;
}

// Mixture function
double sample_mixture(double (*samplers[])(uint64_t*), double* weights, int n_dists, uint64_t* seed)
{
    // Sample from samples with frequency proportional to their weights.
    double sum_weights = array_sum(weights, n_dists);
    double* cumsummed_normalized_weights = (double*)malloc((size_t)n_dists * sizeof(double));
    cumsummed_normalized_weights[0] = weights[0] / sum_weights;
    for (int i = 1; i < n_dists; i++) {
        cumsummed_normalized_weights[i] = cumsummed_normalized_weights[i - 1] + weights[i] / sum_weights;
    }

    double result;
    int result_set_flag = 0;
    double p = sample_uniform(0, 1, seed);
    for (int k = 0; k < n_dists; k++) {
        if (p < cumsummed_normalized_weights[k]) {
            result = samplers[k](seed);
            result_set_flag = 1;
            break;
        }
    }
    if (result_set_flag == 0)
        result = samplers[n_dists - 1](seed);

    free(cumsummed_normalized_weights);
    return result;
}
